db/gorm.go

```diff
@@ -0,0 +1,146 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+var safeNameRE = regexp.MustCompile("^[a-zA-Z0-9_]+$")
+
+// NewMappedSelector returns an object which can be used to automagically query
+// the provided type-mapped database for rows of the parameterized type.
+func NewMappedSelector[T any](executor MappedExecutor) MappedSelector[T] {
+	return &mappedSelector[T]{wrapped: executor}
+}
+
+type mappedSelector[T any] struct {
+	wrapped MappedExecutor
+}
+
+// Query combines the best features of gorp, the go stdlib, and generics.
+// It uses the type parameter of the typeSelector object to automatically look
+// up the proper table name and columns to select. It returns an iterable which
+// yields fully-populated objects of the parameterized type directly. The given
+// clauses MUST be only the bits of a sql query from "WHERE ..." onwards; if
+// they contain any of the "SELECT ... FROM ..." portion of the query it will
+// result in an error. The args take the same kinds of values as gorp's SELECT:
+// either one argument per positional placeholder, or a map of placeholder names
+// to their arguments (https://pkg.go.dev/gopkg.in/gorp.v2#readme-ad-hoc-sql).
+//
+// The caller is responsible for calling `Rows.Close()` when they are done with
+// the query. The caller is also responsible for ensuring that the clauses
+// argument does not contain any user-influenced input.
+func (ts mappedSelector[T]) Query(ctx context.Context, clauses string, args ...interface{}) (Rows[T], error) {
+	// Look up the table to use based on the type of this TypeSelector.
+	var throwaway T
+	tableMap, err := ts.wrapped.TableFor(reflect.TypeOf(throwaway), false)
+	if err != nil {
+		return nil, fmt.Errorf("database model type not mapped to table name: %w", err)
+	}
+
+	return ts.QueryFrom(ctx, tableMap.TableName, clauses, args...)
+}
+
+// QueryFrom is the same as Query, but it additionally takes a table name to
+// select from, rather than automatically computing the table name from gorp's
+// DbMap.
+//
+// The caller is responsible for calling `Rows.Close()` when they are done with
+// the query. The caller is also responsible for ensuring that the clauses
+// argument does not contain any user-influenced input.
+func (ts mappedSelector[T]) QueryFrom(ctx context.Context, tablename string, clauses string, args ...interface{}) (Rows[T], error) {
+	if !safeNameRE.MatchString(tablename) {
+		return nil, fmt.Errorf("unsafe db table name %q", tablename)
+	}
+
+	// We use a very strict mapping of struct fields to table columns here:
+	// - The struct must not have any nested structs, only basic types.
+	// - The struct field names must be case-insensitively identical to the
+	//   column names (no struct tags necessary).
```

Comment:
We should also check there are no two struct fields that map to the same (case-insensitive) column name.

Also - is it true across SQL backends that column names are case-insensitive? If it's not, that's fine; we should just be clear that we're only handling MySQL/MariaDB backends.